### PR TITLE
[WIP] added initial redis cluster transport support code

### PR DESCRIPTION
**Pull request by auvipy**
working on new tests

### CODE REVIEW

File: kombu/transport/redis_cluster.py [hunk 1]
```diff
@@ -0,0 +1,1101 @@
...
+                _, host, port, _, _, _, _ = _parse_url('://' + url)
+
+                hosts.append({
+                    'host': host or '127.0.0.1',
+                    'port': port or DEFAULT_PORT,
+                })
+
+        return hosts, password, path
+
+    def _connparams(self, _r210_options=(
```
Comment (georgepsarakis) [inline kombu/transport/redis_cluster.py]: This tuple should probably be moved to a constant?

Comment (auvipy) [reply kombu/transport/redis_cluster.py]: will check

**Comment by iwinux** [comment]
I've got a weird error when using `SimpleQueue.get()` with this new transport:

```
In [8]: m = q.get(timeout=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-662154a51631> in <module>
----> 1 m = q.get(timeout=1)

/usr/local/lib/python3.7/dist-packages/kombu/simple.py in get(self, block, timeout)
     59                 #    messages are sent over the same socket; also POSIX makes
     60                 #    no guarantees against socket calls returning early.
---> 61                 self.channel.connection.client.drain_events(timeout=remaining)
     62             except socket.timeout:
     63                 raise self.Empty()

/usr/local/lib/python3.7/dist-packages/kombu/connection.py in drain_events(self, **kwargs)
    313             socket.timeout: if the timeout is exceeded.
    314         """
--> 315         return self.transport.drain_events(self.connection, **kwargs)
    316
    317     def maybe_close_channel(self, channel):

/usr/local/lib/python3.7/dist-packages/kombu/transport/virtual/base.py in drain_events(self, connection, timeout)
    961         while 1:
    962             try:
--> 963                 get(self._deliver, timeout=timeout)
    964             except Empty:
    965                 if timeout is not None and monotonic() - time_start >= timeout:

TypeError: get() got multiple values for argument 'timeout'
```

It works fine when using `kombu.transport.redis`.

**Comment by iwinux** [comment]
Seems that `MultiChannelPoller.get` should have signature `get(self, callback, timeout=None)` instead of `get(self, timeout=None)`?

**Comment by auvipy** [comment]
thanks for testing and feedback. I will check and try to push a fix ASAP. keep in mind its work in progress right now

**Comment by iwinux** [comment]
I managed to trim down the code, inherit most stuff from `kombu.transport.redis` and fixed some corner cases (see #1025), resulting in a 250-line module. Some hardcoded things (like `pipeline(True)` in `QoS` forced code duplication, though.

What works so far:

1. `celery worker` seems to work well on Redis Cluster (3 masters + 3 replica) with multi producers + multi consumers.

2. all `celery inspect` is broken for now because the PUBSUB part is not compatible (yet).

3. there're some memory leak issues related to heartbeat (might be PUBSUB's fault too) (see https://github.com/celery/celery/issues/5047).

Thank you for your PR very much. Otherwise I won't even know where to get started. 👍

**Comment by auvipy** [comment]
@iwinux thanks for trying this. If you don't mind could you push your improvements in my branch? as we both are working with it, why should duplicate the effort? lets polish it together? does that sound ok? we will get the credit for this.

**Comment by iwinux** [comment]
That sounds great. I'll send you a pull request :)

**Comment by auvipy** [comment]
amazing! thanks!

**Comment by iwinux** [comment]
The pull request to this pull request is here: https://github.com/auvipy/kombu/pull/2

I've tried to simplify it further but got stuck here. I'm not sure how to implement a new transport *properly* - useful information is buried deep in existing code, obscured by event polling / callbacks, backend specifics and error handling etc.

Also, would it be a good idea to refactor `kombu.transport.redis` so that common code could be shared more cleanly, compared to current hacky method overrides?

**Comment by yakhira** [comment]
guys, i hope you didn't give up with that change :)

**Comment by auvipy** [comment]
this is on list

**Comment by aboganas** [comment]
is this still active?

**Comment by auvipy** [comment]
> is this still active?

can you test this on local machine?

**Comment by NullYing** [comment]
is this still active?

**Comment by auvipy** [comment]
I removed my personal fork so this need copy-paste if someone wants to try & contribute

**Comment by ukm21** [comment]
Has anyone fixed this issue

File "/usr/local/lib/python3.6/site-packages/celery/worker/worker.py", line 205, in start
self.blueprint.start(self)
File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
step.start(parent)
File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 369, in start
return self.obj.start()
File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 318, in start
blueprint.start(self)
File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
step.start(parent)
File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 596, in start
c.loop(*c.loop_args())
File "/usr/local/lib/python3.6/site-packages/celery/worker/loops.py", line 113, in synloop
connection.drain_events(timeout=2.0)
File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 323, in drain_events
return self.transport.drain_events(self.connection, **kwargs)
File "/usr/local/lib/python3.6/site-packages/kombu/transport/virtual/base.py", line 963, in drain_events
get(self._deliver, timeout=timeout)
TypeError: get() got multiple values for argument 'timeout'

**Comment by jeffreybrowning** [comment]
New comment for people to indicate support.

**Comment by dligthart** [comment]
is this still active ? would be really nice to have.

**Comment by jeffreybrowning** [comment]
@dligthart In my experience, the celery team is stretched pretty thin and @auvipy here is suggesting someone jump in and solve contribute to the project if they want this done.

**Comment by auvipy** [comment]
i might revisit it in near future If no one take this over

**Comment by aboganas** [comment]
Is there a way to financially contribute to such a feature?

**Comment by auvipy** [comment]
> Is there a way to financially contribute to such a feature?

you can reach me out at: auvipy@gmail.com

**Comment by pabclsn** [comment]
any news on this one ;) ?

**Comment by haobibo** [comment]
@auvipy  I see this FR added milestone 5.1.0 and 6.0. Is it implemented somewhere else?
This will be a really useful feature for many use cases.

**Comment by auvipy** [comment]
> @auvipy I see this FR added milestone 5.1.0 and 6.0. Is it implemented somewhere else? This will be a really useful feature for many use cases.

this will be moved to somewhere else for now

**Comment by brunotacca** [comment]
Is this in progress or being tracked somewhere else in this repo?

Thanks

**Comment by zs-neo** [comment]
@auvipy Is the work still ongoing? Celery is very important to our company, and we have found that it does not support Redis cluster. We want to support it on Kombu, does this issue still accept PR?

**Comment by auvipy** [comment]
Yes it does

**Comment by LeoNumber1** [comment]
> @auvipy Is the work still ongoing? Celery is very important to our company, and we have found that it does not support Redis cluster. We want to support it on Kombu, does this issue still accept PR?

@zs-neo hello, my project just need Celery work in redis cluster. Can we have a conversation？Thanks。
Email：liyuan.go@bytedance.com

**Comment by zzJinux** [comment]
https://github.com/celery/kombu/blob/63108c292c7a5e380baa45b77f9a8b7a735c448d/kombu/transport/redis_cluster.py#L152-L154

Is priority queue not supported for redis cluster transport?

**Comment by Ragnarow** [comment]
Hello, can the MR be merged for next release?

**Comment by auvipy** [comment]
that PR is scheduled for v5.6

**Pull request closed without merging**